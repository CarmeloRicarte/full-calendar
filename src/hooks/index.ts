export * from "./dispatchSelector";
export * from "./useCalendarStore";
export * from "./useForm";
export * from "./useUiStore";
