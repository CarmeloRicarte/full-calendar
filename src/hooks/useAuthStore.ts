import { useSelector } from "react-redux";
import { calendarApi } from "../api";
import {
  clearErrorMessage,
  onChecking,
  onLogin,
  onLogout,
  RootState,
} from "../store";
import { useAppDispatch } from "./dispatchSelector";

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector(
    (state: RootState) => state.auth
  );
  const dispatch = useAppDispatch();

  const startLogin = async ({
    email,
    password,
  }: {
    email: string;
    password: string;
  }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth", {
        email,
        password,
      });
      setTokenLocalStorage(data.token);

      dispatch(
        onLogin({
          name: data.name,
          uid: data.uid,
        })
      );
    } catch (error) {
      dispatch(onLogout("Credenciales incorrectas"));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 1000);
    }
  };

  const startRegister = async ({
    name,
    email,
    password,
  }: {
    name: string;
    email: string;
    password: string;
  }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth/new", {
        name,
        email,
        password,
      });
      setTokenLocalStorage(data.token);

      dispatch(
        onLogin({
          name: data.name,
          uid: data.uid,
        })
      );
    } catch (error: any) {
      if (error.response.data?.msg) {
        dispatch(onLogout(error.response.data?.msg));
      } else {
        const keyFirstElement = Object.keys(error.response.data?.errors)[0];
        keyFirstElement &&
          dispatch(onLogout(error.response.data?.errors[keyFirstElement]?.msg));
      }

      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 1000);
    }
  };

  const setTokenLocalStorage = (token: string) => {
    localStorage.setItem("token", token);
    localStorage.setItem(
      "token-expires-date",
      new Date().getTime() + 2 * 60 * 60 * 1000 + ""
    );
  };

  const checkAuthToken = async () => {
    const token = localStorage.getItem("token");
    if (!token) return dispatch(onLogout());
    try {
      const { data } = await calendarApi.get("/auth/renew");

      // logout user if token expired
      const expirationTokenDate = localStorage.getItem("token-expires-date");
      if (expirationTokenDate && +expirationTokenDate < new Date().getTime()) {
        dispatch(onLogout());
      } else {
        setTokenLocalStorage(data);
        dispatch(
          onLogin({
            name: data.name,
            uid: data.uid,
          })
        );
      }
    } catch (error) {
      localStorage.clear();
      dispatch(onLogout());
    }
  };

  return {
    //* Properties
    status,
    user,
    errorMessage,
    //* Methods
    startLogin,
    startRegister,
    checkAuthToken,
  };
};
