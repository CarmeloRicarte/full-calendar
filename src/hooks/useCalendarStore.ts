import { useAppDispatch } from "./dispatchSelector";
import { useSelector } from "react-redux";
import { onSetActiveEvent, RootState } from "../store";
import { CalendarEvent } from "../calendar/interfaces";

export const useCalendarStore = () => {
  const dispatch = useAppDispatch();
  const { events, activeEvent } = useSelector(
    (state: RootState) => state.calendar
  );

  const setActiveEvent = (calendarEvent: CalendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  return {
    //* Properties
    events,
    activeEvent,
    //*Methods
    setActiveEvent,
  };
};
