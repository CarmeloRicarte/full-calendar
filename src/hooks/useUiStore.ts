import { useSelector } from "react-redux";
import { onCloseDateModal, onOpenDateModal, RootState } from "../store";
import { useAppDispatch } from "./dispatchSelector";

export const useUiStore = () => {
  const dispatch = useAppDispatch();
  const { isDateModalOpen } = useSelector((state: RootState) => state.ui);

  const openDateModal = () => {
    dispatch(onOpenDateModal());
  };

  const closeDateModal = () => dispatch(onCloseDateModal());

  return {
    //* Properties
    isDateModalOpen,

    //*Methods
    openDateModal,
    closeDateModal,
  };
};
