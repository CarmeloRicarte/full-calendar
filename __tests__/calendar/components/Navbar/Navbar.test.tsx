import { describe, test, expect, vi, beforeEach } from "vitest";
import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { Navbar } from "../../../../src/calendar/components/Navbar/Navbar";
import * as useAuthStore from "../../../../src/hooks/useAuthStore";
import { testUserCredentials } from "../../../__fixtures__/testUser";

vi.mock("../../../../src/hooks/useAuthStore");

describe("Tests of <Navbar />", () => {
  const mockStartLogout = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  test("should render correctly showing the user name", () => {
    (useAuthStore as any).useAuthStore.mockReturnValue({
      user: testUserCredentials,
      startLogout: mockStartLogout,
    });
    const { container } = render(<Navbar />);
    expect(screen.getByText(testUserCredentials.name)).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  test("should call startLogout on click button", () => {
    (useAuthStore as any).useAuthStore.mockReturnValue({
      user: testUserCredentials,
      startLogout: mockStartLogout,
    });
    render(<Navbar />);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(mockStartLogout).toHaveBeenCalled();
  });
});
