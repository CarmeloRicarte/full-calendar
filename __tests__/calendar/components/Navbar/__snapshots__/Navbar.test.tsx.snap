// Vitest Snapshot v1

exports[`Tests of <Navbar /> > should render correctly showing the user name 1`] = `
<div>
  <div
    class="navbar navbar-dark bg-dark mb-4 px-4"
  >
    <span
      class="navbar-brand"
    >
      <i
        class="fas fa-calendar-alt"
      />
      Â  
      Test User
    </span>
    <button
      class="btn btn-outline-danger"
      type="button"
    >
      <i
        class="fas fa-sign-out-alt"
      />
      <span>
        Salir
      </span>
    </button>
  </div>
</div>
`;
